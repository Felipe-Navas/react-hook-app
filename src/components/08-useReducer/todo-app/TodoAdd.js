import React from 'react';
import { useForm } from '../../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, reset] = useForm({
    description: '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Add TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          className="form-control"
          type="text"
          name="description"
          placeholder="Enter a todo description"
          autoComplete="off"
          onChange={handleInputChange}
          value={description}
        />
        <div className="d-grid gap-2">
          <button className="btn btn-success mt-1" type="submit">
            Add Todo
          </button>
        </div>
      </form>
    </>
  );
};
