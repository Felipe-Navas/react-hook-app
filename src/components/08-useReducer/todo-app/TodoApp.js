import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../../hooks/useForm';
import { TodoList } from './TodoList';
import './styles.css';

const init = () => {
  // return [
  //   {
  //     id: new Date().getTime(),
  //     desc: 'Deploy an application',
  //     done: false,
  //   },
  // ];

  return JSON.parse(localStorage.getItem('Todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const [{ description }, handleInputChange, reset] = useForm({
    description: '',
  });

  useEffect(() => {
    localStorage.setItem('Todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    const action = {
      type: 'add',
      payload: newTodo,
    };
    dispatch(action);
    reset();
  };

  return (
    <div>
      <h1>TodoApp ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <h4>Add TODO</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              className="form-control"
              type="text"
              name="description"
              placeholder="Enter a todo description"
              autoComplete="off"
              onChange={handleInputChange}
              value={description}
            />
            <div className="d-grid gap-2">
              <button className="btn btn-success mt-1" type="submit">
                Add Todo
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
