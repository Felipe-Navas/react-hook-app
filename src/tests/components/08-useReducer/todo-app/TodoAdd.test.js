import React from 'react';
import { shallow } from 'enzyme';
import { TodoAdd } from '../../../../components/08-useReducer/todo-app/TodoAdd';

describe('Testing the <TodoAdd /> component', () => {
  const handleAddTodo = jest.fn();
  const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should not call the handleAddTodo function', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });
    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('should call the handleAddTodo function with the parameter', () => {
    const value = 'Aprender a programar';
    wrapper
      .find('input')
      .simulate('change', { target: { value, name: 'description' } });
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });
    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAddTodo).toHaveBeenCalledWith({
      desc: value,
      done: false,
      id: expect.any(Number),
    });
    expect(wrapper.find('input').prop('value')).toBe('');
  });
});
