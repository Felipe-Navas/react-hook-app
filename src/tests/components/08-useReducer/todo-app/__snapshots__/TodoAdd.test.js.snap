// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the <TodoAdd /> component should render correctly 1`] = `
<Fragment>
  <h4>
    Add TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Enter a todo description"
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-success mt-1"
        type="submit"
      >
        Add Todo
      </button>
    </div>
  </form>
</Fragment>
`;
